text/0000-regex-1.0.md

```diff
@@ -0,0 +1,980 @@
+3. Any change that causes a valid regex pattern to change its matching
+   semantics *is* a breaking change. (For example, changing `\b` from "word
+   boundary assertion" to "backspace character.")
+
+Bug fixes are exceptions to both (2) and (3).
+
+Another interesting exception to (2) is that compiling a regex can fail if the
+entire compiled object would exceed some pre-defined user configurable size.
+In particular, future changes to the compiler could cause certain instructions
+to use more memory, or indeed, the representation of the compiled regex could
+change completely. This could cause a regex that fit under the size limit to
+no longer fit, and therefore fail to compile. These cases are expected to be
+extremely rare in practice. Notably, the default size limit is `10MB`.
+
+### Concrete syntax
+[concrete-syntax]: #concrete-syntax
+
+The syntax is exhaustively documented in the current public API documentation:
+http://doc.rust-lang.org/regex/regex/index.html#syntax
+
+To my knowledge, the evolution as proposed in this RFC has been followed since
+`regex` was created. The syntax has largely remain unchanged with few
+additions.
+
+### Expansion concerns
+[expansion-concerns]: #expansion-concerns
+
+There are a few possible avenues for expansion, and we take measures to make
+sure they are possible with respect to API evolution.
+
+* Escape sequences are often blessed with special semantics. For example, `\d`
+  is a Unicode character class that matches any digit and `\b` is a word
+  boundary assertion. We may one day like to add more escape sequences with
+  special semantics. For this reason, any unrecognized escape sequence makes a
+  pattern invalid.
+* If we wanted to expand the syntax with various look-around operators, then it
+  would be possible since most common syntax is considered an invalid pattern
+  today. In particular, all of the [syntactic forms listed
+  here](http://www.regular-expressions.info/refadv.html) are invalid patterns
+  in `regex`.
+* Character class sets are another potentially useful feature that may be worth
+  adding. Currently, [various forms of set
+  notation](http://www.regular-expressions.info/refcharclass.html) are treated
+  as valid patterns, but this RFC proposes making them invalid patterns before
+  `1.0`.
+* Additional named Unicode classes or codepoints may be desirable to add.
+  Today, any pattern of the form `\p{NAME}` where `NAME` is unrecognized is
+  considered invalid, which leaves room for expansion.
+
+## Core API
+[core-api]: #core-api
+
+The core API of the `regex` crate is the `Regex` type:
+
+```rust
+pub struct Regex(_);
+```
+
+It has one primary constructor:
+
+```rust
+impl Regex {
+  /// Creates a new regular expression. If the pattern is invalid or otherwise
+  /// fails to compile, this returns an error.
+  pub fn new(pattern: &str) -> Result<Regex, Error>;
+}
+```
+
+And five core search methods. All searching completes in worst case linear
+time.
```

Comment:
(Not directly relevant, but is this linear with respect to both `pattern` and `text`, i.e., O(n + m)? Or maybe linear with respect to each while holding the other constant, i.e., O(nm)?)